src/components/feedback/Dialog.tsx

```diff
@@ -48,46 +49,77 @@ type ComponentProps = {
    * Optional TransitionComponent for open (mount) and close (unmount) transitions
    */
   transitionComponent?: TransitionComponent;
-
-  /**
-   * Dialog lays out its content in a Panel by default, but `custom` allows any
-   * layout of Dialog children
-   */
-  variant?: 'panel' | 'custom';
 };
 
 // This component forwards a number of props on to `Panel` but always sets the
 // `fullWidthHeader` prop to `true`.
-export type DialogProps = PresentationalProps &
+export type PanelDialogProps = PresentationalProps &
   ComponentProps &
-  Omit<PanelProps, 'fullWidthHeader'>;
+  Omit<PanelProps, 'fullWidthHeader'> & {
+    /** Dialog lays out its content in a Panel by default */
+    variant?: 'panel';
+  };
+
+type NonPanelDialogProps = PresentationalProps &
+  ComponentProps & {
+    /** `custom` allows any layout of Dialog children */
+    variant: 'custom';
+
+    /** @deprecated Title is ignored in custom dialogs */
+    title?: string;
+  };
+
+export type DialogProps = PanelDialogProps | NonPanelDialogProps;
+
+export function isPanelDialogProps(
+  props: DialogProps,
+): props is PanelDialogProps {
+  return props.variant !== 'custom';
+}
+
+function isPanelProps(
+  props: DialogProps,
+): props is Omit<PanelProps, 'fullWidthHeader'> {
+  return 'title' in props;
+}
 
 /**
  * Show a dialog
  */
-export default function Dialog({
-  closeOnClickAway = false,
-  closeOnEscape = false,
-  closeOnFocusAway = false,
-  children,
-  initialFocus = 'auto',
-  restoreFocus = false,
-  transitionComponent: TransitionComponent,
-  variant = 'panel',
-
-  classes,
-  elementRef,
-
-  // Forwarded to Panel
-  buttons,
-  icon,
-  onClose,
-  paddingSize = 'md',
-  scrollable = true,
-  title,
-
-  ...htmlAttributes
-}: DialogProps) {
+export default function Dialog(props: DialogProps) {
+  const {
+    closeOnClickAway = false,
+    closeOnEscape = false,
+    closeOnFocusAway = false,
+    children,
+    initialFocus = 'auto',
+    restoreFocus = false,
+    transitionComponent: TransitionComponent,
+
+    classes,
+    elementRef,
+    onClose,
+
+    ...rest
+  } = props;
+  const {
+    buttons,
+    icon,
+    paddingSize = 'md',
+    title = '',
+    scrollable = true,
+    ...htmlAttributes
+  } = isPanelProps(rest)
+    ? rest
+    : {
+        buttons: undefined,
+        icon: undefined,
+        paddingSize: undefined,
+        title: undefined,
+        scrollable: undefined,
+        ...rest,
+      };
```

Comment:
This makes sure the panel props are stripped out of the `htmlAttributes`, avoiding type check errors when passing them to the `div` element.